tests/Application/Persistence/RetrySafeEntityManagerTest.php

```diff
@@ -152,6 +152,65 @@ public function testFlushWithRetry(): void
         $this->retrySafeEntityManager->flush();
     }
 
+    public function testRefresh(): void
+    {
+        $underlyingEmProphecy = $this->prophesize(EntityManager::class);
+        $underlyingEmProphecy
+            ->refresh(Argument::type(Donation::class),LockMode::PESSIMISTIC_WRITE)
+            ->shouldBeCalledOnce();
+
+        $retrySafeEntityManagerReflected = new ReflectionClass($this->retrySafeEntityManager);
+
+        $emProperty = $retrySafeEntityManagerReflected->getProperty('entityManager');
+        $emProperty->setValue($this->retrySafeEntityManager, $underlyingEmProphecy->reveal());
+
+        $app = $this->getAppInstance();
+        /** @var Container $container */
+        $container = $app->getContainer();
+        $container->set(EntityManager::class, $underlyingEmProphecy->reveal());
+
+        $this->retrySafeEntityManager->refresh(new Donation(), LockMode::PESSIMISTIC_WRITE);
+    }
+
+    public function testRefreshWithRetry(): void
+    {
+        $app = $this->getAppInstance();
+        /** @var Container $container */
+        $container = $app->getContainer();
+
+        // First underlying EM should throw closed error.
+        $underlyingEmProphecy1 = $this->prophesize(EntityManager::class);
+        $underlyingEmProphecy1
+            ->refresh(Argument::type(Donation::class), LockMode::PESSIMISTIC_WRITE)
+            ->willThrow(new EntityManagerClosed())
+            ->shouldBeCalledOnce();
+
+        // Second should work.
+        $underlyingEmProphecy2 = $this->prophesize(EntityManager::class);
+        $underlyingEmProphecy2
+            ->refresh(Argument::type(Donation::class), LockMode::PESSIMISTIC_WRITE)
+            ->shouldBeCalledOnce();
+
+        /** @var array<string, mixed> $settings */
+        $settings = $container->get('settings');
+        /** @var array<string, mixed> $doctrineSettings */
+        $doctrineSettings = $settings['doctrine'];
+        /** @var array<string, mixed> $connectionSettings */
+        $connectionSettings = $doctrineSettings['connection'];
+        $this->retrySafeEntityManager = $this->getRetrySafeEntityManagerPartialMock(
+            $this->getConfiguration($container),
+            $connectionSettings,
+            $underlyingEmProphecy1->reveal(),
+            $underlyingEmProphecy2->reveal(),
+        );
+
+        $container->set(EntityManager::class, $underlyingEmProphecy1->reveal());
+        $container->set(RetrySafeEntityManager::class, $this->retrySafeEntityManager);
+        $container->set(EntityManagerInterface::class, $this->retrySafeEntityManager);
+
+        $this->retrySafeEntityManager->refresh(new Donation(), LockMode::PESSIMISTIC_WRITE);
+    }
```

Comment:
I simplified this test quite a lot - I think it still this version still tests what you want to test, and is easier to understand and less coupled to various things. Let me know what you think?

```suggestion
    public function testRefreshRetriesOnEmClosed(): void
    {
        // First underlying EM should throw closed error.
        $underlyingEmProphecy1 = $this->prophesize(EntityManager::class);
        $underlyingEmProphecy1
            ->refresh(Argument::type(Donation::class), LockMode::PESSIMISTIC_WRITE)
            ->willThrow(new EntityManagerClosed());

        // Second should work.
        $underlyingEmProphecy2 = $this->prophesize(EntityManager::class);
        $underlyingEmProphecy2
            ->refresh(Argument::type(Donation::class), LockMode::PESSIMISTIC_WRITE)
            ->shouldBeCalledOnce();

        $container = $this->getAppInstance()->getContainer();
        assert($container instanceof Container);
        $retrySafeEntityManager = $this->getRetrySafeEntityManagerPartialMock(
            $this->getConfiguration($container),
            ['driver' => 'pdo_mysql'],
            $underlyingEmProphecy1->reveal(),
            $underlyingEmProphecy2->reveal(),
        );

        $retrySafeEntityManager->refresh(new Donation(), LockMode::PESSIMISTIC_WRITE);
    }
```

Reply:
Much better, thanks!